plugins/debuginfo-install.py

```diff
@@ -74,59 +76,170 @@ def configure(self):
         self.base.repos.enable_debug_repos()
 
     def run(self):
-        self.packages = self.base.sack.query()
-        self.packages_available = self.packages.available()
         errors_spec = []
 
+        debuginfo_suffix_len = len(Package.DEBUGINFO_SUFFIX)
+        debugsource_suffix_len = len(Package.DEBUGSOURCE_SUFFIX)
+
         for pkgspec in self.opts.package:
-            package_query = sorted(dnf.subject.Subject(pkgspec).get_best_query(
-                self.cli.base.sack).filter(arch__neq='src'), reverse=True)
-            if not package_query:
-                msg = _('No match for argument: %s')
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
+
+            query = solution["query"].filterm(arch__neq='src')
+
+            package_dict = query.available()._name_dict()
+            # installed versions of packages have priority, replace / add them to the dict
+            package_dict.update(query.installed()._name_dict())
+
+            if not package_dict:
+                logger.info(_('No match for argument: %s'), self.base.output.term.bold(pkgspec))
                 errors_spec.append(pkgspec)
-            for pkg in package_query:
-                self._di_install(pkg)
+                continue
 
-        if not self.dbgdone:
-            logger.info(_("No debuginfo packages available to install"))
+            # Remove debuginfo packages if their base packages are in the query.
+            # They can get there through globs and they break the installation
+            # of debug packages with the same version as the installed base
+            # packages. If the base package of a debuginfo package is not in
+            # the query, the user specified a debug package on the command
+            # line. We don't want to ignore those, so we will install them.
+            # But, in this case the version will not be matched to the
+            # installed version of the base package, as that would require
+            # another query and is further complicated if the user specifies a
+            # version themselves etc.
+            for name in list(package_dict.keys()):
+                if name.endswith(Package.DEBUGINFO_SUFFIX):
+                    if name[:-debuginfo_suffix_len] in package_dict:
+                        package_dict.pop(name)
+                if name.endswith(Package.DEBUGSOURCE_SUFFIX):
+                    if name[:-debugsource_suffix_len] in package_dict:
+                        package_dict.pop(name)
```

Comment:
I believe that after the latest changes `for name in list(package_dict.keys()):` loop could be removed completely. If we will provide all possibility to solver (list of all packages and not only the latest) there is no need to reduce data set.

Reply:
This won't work for the case when there is e.g. package `bar-2.0` installed and `bar-{debuginfo,debugsource}-2.0` don't exist. Expected behavior is no `bar-{debuginfo,debugsource}` aren't installed, but if I remove this piece of code, it will install the latest versions of `bar-{debuginfo,debugsource}` available.

There is a testcase with this scenario (called _"multiple packages, a glob that doesn't find debuginfo for the installed version"_) in the CI PR, which breaks if I try this (even if I pass all the `-debuginfo` packages to the solver instead of just the first one).

Reply:
Then it must be resolved somewhere else.

Reply:
> This won't work for the case when there is e.g. package `bar-2.0` installed and `bar-{debuginfo,debugsource}-2.0` don't exist. Expected behavior is no `bar-{debuginfo,debugsource}` are installed, but if I remove this piece of code, it will install the latest versions of `bar-{debuginfo,debugsource}` available.

In this case I believe that it should fail.

> 
> There is a testcase with this scenario (called _"multiple packages, a glob that doesn't find debuginfo for the installed version"_) in the CI PR, which breaks if I try this (even if I pass all the `-debuginfo` packages to the solver instead of just the first one).

Reply:
Please provide detailed example including set packages installed, available and argument used for command.

Reply:
The referenced test case is just that...